### PR TITLE
Avoid spurious ambiguity warning with inherited overloaded symbols

### LINKED ISSUE #12850: Spurious "ambiguous reference inherited vs outer" warning
```scala
trait T {
  def pm(x: Int) = 1
  def pm(x: String) = 2
}

package object p extends T {
  def t = pm(1)
}
```

Gives a spurious warning

```
Test.scala:7: warning: reference to pm is ambiguous;
it is both defined in the enclosing trait T and inherited in the enclosing package object p as method pm (defined in trait T)
In Scala 2, symbols inherited from a superclass shadow symbols defined in an outer scope.
Such references are ambiguous in Scala 3. To continue using the inherited symbol, write `this.pm`.
Or use `-Wconf:msg=legacy-binding:s` to silence this warning. [quickfixable]
  def t = pm(1)
          ^
```

### PR DESCRIPTION
**Pull request by lrytz**
Follow-up for https://github.com/scala/scala/pull/10321 (which is a follow-up for https://github.com/scala/scala/pull/10220)

Fixes https://github.com/scala/bug/issues/12850

### CODE REVIEW

**Review by SethTisue — APPROVED** [review]
this is primarily a "I'm okay with this going into 2.13.12" review. as in "we don't need to reset the community testing clock". let's still see what Som says

File: src/compiler/scala/tools/nsc/typechecker/Contexts.scala [hunk 1]
```diff
@@ -1637,7 +1637,11 @@ trait Contexts { self: Analyzer =>
           done = (cx eq NoContext) || foundCompetingSymbol()
           if (!done && (cx ne NoContext)) cx = cx.outer
         }
-        if (defSym.exists && (defSym ne defSym0)) {
+        val nonOverlapping = defSym.exists && {
+          if (defSym.isOverloaded || defSym0.isOverloaded) defSym.alternatives.intersect(defSym0.alternatives).isEmpty
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala]:
```suggestion
          if (defSym.isOverloaded || defSym0.isOverloaded) !defSym.alternatives.exists(defSym0.alternatives.contains)
```

**Review by som-snytt — APPROVED** [review]
Small suggestion, as `intersect` is heavyweight.

I started looking for other problematic comparisons; I expected them everywhere. One may be in `nextDefinition`, where `memberType` is overload-aware, then we test `last == defSym`.

**Comment by som-snytt** [comment]
Maybe there will be a follow-up in 2.13.13.

It would be nice to have `sym.isDistinctFrom(other)`. Still better than `sym.filter(other.alternatives.contains) == NoSymbol`. I did not know `sym.suchThat` asserts that the result is not overloaded.

**Pull request merged**